builtin/providers/test/resource_with_custom_diff.go

```diff
@@ -0,0 +1,84 @@
+package test
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func testResourceCustomDiff() *schema.Resource {
+	return &schema.Resource{
+		Create:        testResourceCustomDiffCreate,
+		Read:          testResourceCustomDiffRead,
+		CustomizeDiff: testResourceCustomDiffCustomizeDiff,
+		Update:        testResourceCustomDiffUpdate,
+		Delete:        testResourceCustomDiffDelete,
+		Schema: map[string]*schema.Schema{
+			"required": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"computed": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"index": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"veto": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func testResourceCustomDiffCreate(d *schema.ResourceData, meta interface{}) error {
+	d.SetId("testId")
+
+	// Required must make it through to Create
+	if _, ok := d.GetOk("required"); !ok {
+		return fmt.Errorf("missing attribute 'required', but it's required")
+	}
+
+	_, new := d.GetChange("computed")
+	expected := new.(int) - 1
+	actual := d.Get("index").(int)
+	if expected != actual {
+		return fmt.Errorf("expected computed to be 1 ahead of index, got computed: %d, index: %d", expected, actual)
+	}
+	d.Set("index", new)
+
+	return testResourceCustomDiffRead(d, meta)
+}
+
+func testResourceCustomDiffRead(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func testResourceCustomDiffCustomizeDiff(d *schema.ResourceDiff, meta interface{}) error {
+	if d.Get("veto").(bool) == true {
+		return fmt.Errorf("veto is true, diff vetoed")
+	}
+	// Note that this gets put into state after the update, regardless of whether
+	// or not anything is acted upon in the diff.
+	d.SetNew("computed", d.Get("computed").(int)+1)
```

Comment:
From what I understand, I don't think there should be any ill effects. What would happen is any values that got set in with `Set*` functions in the `schema.ResourceData` that gets passed along would be set in the `set` writer layer, while the data set in the diff function would have been already properly rolled into the diff and would exist exclusively in the `diff` reader layer. `set` takes precedence, so what I'm pretty sure will happen is that after `Create` or `Update` returns any value that has been put into the `set` writer will be the one that `wins` when the new state is calculated.

The final step of a `Resource.Apply` is to [merge the state](https://github.com/hashicorp/terraform/blob/master/helper/schema/resource.go#L195) using [`ResourceData.State()`](https://github.com/hashicorp/terraform/blob/master/helper/schema/resource_data.go#L244), which reconciles the state with all the data collected in the various levels of the writer, starting with the `set` layer (or `state` if in partial mode without a `SetPartial` for the specific key, which bypasses the diff anyway, before this change), and returns it, which ultimately [becomes the new state](https://github.com/hashicorp/terraform/blob/master/terraform/eval_apply.go#L79).

The good old "diffs didn't match during apply" error was something I was worried about too, but I don't think we are at risk for that, mainly because [`EvalCompareDiff` gets evaluated before `EvalApply`](https://github.com/hashicorp/terraform/blob/master/terraform/node_resource_apply.go#L279), which means that 1) state hasn't been written yet, and 2) `Create` or `Update` will not have fired anyway, meaning that there is no risk in either of these creating an issue. What I think we need to be more worried about is a pre-written plan drifting too far from the current *real* state of the resource that any custom diff behaviour is pulling to calculate the diff. However, I think that's an accurate case of what you were describing in #8769 where the user may need to re-refresh (and then re-plan, ultimately).

I think this could be properly vetoed - the drift could be detected via checking the current real world value and then comparing it to the data from `GetChange`, giving the opportunity to provide a friendly error to the user to request they re-plan.